File: Java8-JavaFX/src/sevenguis/circledrawer/CircleDrawerReactFX.java
package sevenguis.circledrawer;

import javafx.application.Application;
import javafx.event.ActionEvent;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.Slider;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.stage.Popup;
import javafx.stage.Stage;
import org.reactfx.EventStream;
import sevenguis.circledrawer.undo.UndoManager;
import sevenguis.circledrawer.undo.UndoableEdit;

import java.util.ArrayList;
import java.util.List;

import static org.reactfx.EventStreams.*;

// https://gist.github.com/TomasMikula/b04dd89da1584597fa14
public class CircleDrawerReactFX extends Application {

    public void start(Stage stage) {
        Button undo = new Button("Undo");
        Button redo = new Button("Redo");
        CircleDrawerCanvasReactFX canvas = new CircleDrawerCanvasReactFX();

        undo.setOnAction(e -> canvas.undo());
        redo.setOnAction(e -> canvas.redo());

        BorderPane root = new BorderPane();
        root.setPadding(new Insets(10));
        HBox top = new HBox(10, undo, redo);
        top.setPadding(new Insets(0, 0, 10, 0));
        root.setTop(top);
        root.setCenter(canvas);

        stage.setScene(new Scene(root));
        stage.setTitle("Circle Drawer ReactFX");
        stage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}

class CircleDrawerCanvasReactFX extends Canvas {

    private List<Circle> circles;
    private UndoManager undoManager;

    public CircleDrawerCanvasReactFX() {
        super(400, 400);
        circles = new ArrayList<>();
        undoManager = new UndoManager();

        Button diameterEntry = new Button("Diameter...");
        Popup popup = new Popup();
        popup.getContent().add(diameterEntry);

        EventStream<MouseEvent> leftPressesToVoid = eventsOf(this, MouseEvent.MOUSE_PRESSED)
                .filter(MouseEvent::isPrimaryButtonDown)
                .filter(e -> getNearestCircleAt(e.getX(), e.getY()) == null);
        EventStream<Circle> addedCircles = leftPressesToVoid
                .map(e -> new Circle((int)e.getX(), (int)e.getY()))
                .hook(this::addCircle);
        EventStream<Circle> hoveredCircles = eventsOf(this, MouseEvent.MOUSE_MOVED)
                .map(e -> getNearestCircleAt(e.getX(), e.getY()));
        merge(addedCircles, hoveredCircles).subscribe(this::draw);

        eventsOf(this, MouseEvent.MOUSE_PRESSED).subscribe(e -> { if (popup.isShowing()) popup.hide(); });
        EventStream<MouseEvent> rightPressesToCircle = eventsOf(this, MouseEvent.MOUSE_PRESSED)
                .filter(MouseEvent::isPopupTrigger)
                .filter(e -> getNearestCircleAt(e.getX(), e.getY()) != null);
        rightPressesToCircle.subscribe(e -> popup.show(this, e.getScreenX(), e.getScreenY()));
        EventStream<Circle> selectedCircles = rightPressesToCircle.map(e -> getNearestCircleAt(e.getX(), e.getY()));
        EventStream<ActionEvent> diameterEntryClicks = eventsOf(diameterEntry, ActionEvent.ACTION);
        selectedCircles.emitOn(diameterEntryClicks).subscribe(c -> {
            popup.hide();
            showDialog(c);
        });

        draw();
    }

    void draw() { draw(null); }

    void draw(Circle hovered) {
        GraphicsContext g = getGraphicsContext2D();
        g.setFill(Color.WHITE);
        g.setStroke(Color.BLACK);
        g.fillRect(0, 0, getWidth(), getHeight());
        g.strokeRect(0, 0, getWidth(), getHeight());

        g.setFill(Color.LIGHTGRAY);
        for (Circle c : circles) {
            int offset = c.getDiameter() / 2;
            if (c.equals(hovered)) {
                g.fillOval(c.getX() - offset, c.getY() - offset, c.getDiameter(), c.getDiameter());
            }
            g.strokeOval(c.getX() - offset, c.getY() - offset, c.getDiameter(), c.getDiameter());
        }
    }

    void addCircle(Circle circle) {
        undoManager.addEdit(new CreateCircleEdit(circle));
        circles.add(circle);
    }

    public Circle getNearestCircleAt(double x, double y) {
        Circle circle = null;
        double minDist = Double.MAX_VALUE;
        for (Circle c : circles) {
            double d = Math.sqrt(Math.pow(x - c.getX(), 2) + Math.pow(y - c.getY(), 2));
            if (d <= c.getDiameter()/2 && d < minDist) {
                circle = c;
                minDist = d;
            }
        }
        return circle;
    }

    void showDialog(Circle selected) {
        Stage dialog = new Stage();
        Label info = new Label(String.format("Adjust diameter of circle at (%s, %s)",
                selected.getX(), selected.getY()));
        Slider slider = new Slider(10, 50, selected.getDiameter());
        VBox content = new VBox(10, info, slider);
        content.setPadding(new Insets(10));
        dialog.setScene(new Scene(content));
        dialog.show();

        slider.valueProperty().addListener((v, o, n) -> {
            selected.setDiameter(n.intValue());
            draw();
        });
        int oldDiameter = selected.getDiameter();
        dialog.setOnCloseRequest(e -> undoManager.addEdit(
                new ChangeDiameterEdit(selected, oldDiameter, selected.getDiameter())));
    }

    public void undo() {
        undoManager.undo();
        draw();
    }

    public void redo() {
        undoManager.redo();
        draw();
    }

    private class CreateCircleEdit extends UndoableEdit {

        private Circle circle;

        CreateCircleEdit(Circle circle) {
            this.circle = circle;
        }

        public void undo() {
            circles.remove(circle);
        }

        public void redo() {
            circles.add(circle);
        }

    }

    private class ChangeDiameterEdit extends UndoableEdit {

        private Circle circle;
        private int oldDiameter, newDiameter;

        ChangeDiameterEdit(Circle circle, int oldDiameter, int newDiameter) {
            this.circle = circle;
            this.oldDiameter = oldDiameter;
            this.newDiameter = newDiameter;
        }

        public void undo() {
            circle.setDiameter(oldDiameter);
        }

        public void redo() {
            circle.setDiameter(newDiameter);
        }

    }

}

File: Scala-ScalaFX/src/main/scala/sevenguis/circledrawer/CircleDrawerReactFX.scala
package sevenguis.circledrawer


import scalafx.application.JFXApp
import scalafx.application.JFXApp.PrimaryStage
import scalafx.scene.Scene
import scalafx.scene.control.{Slider, Label, Button}
import scalafx.scene.layout.{VBox, BorderPane, HBox}
import scalafx.geometry.Insets
import scalafx.scene.canvas.Canvas
import scalafx.scene.paint.Color
import scalafx.event.ActionEvent
import scalafx.Includes._
import scala.math._
import scalafx.stage.{Popup, WindowEvent, Stage}
import javafx.beans.value.ObservableValue

import org.reactfx.EventStream
import org.reactfx.EventStreams._
import sevenguis.ReactFXIntegration._
import sevenguis.Scala2Java8._

// https://gist.github.com/TomasMikula/b04dd89da1584597fa14
object CircleDrawerFX extends JFXApp {
  val undo = new Button("Undo")
  val redo = new Button("Redo")
  val canvas = new CircleDrawerCanvasFX()

  undo.onAction = (event: ActionEvent) => canvas.undo()
  redo.onAction = (event: ActionEvent) => canvas.redo()

  stage = new PrimaryStage {
    title = "Circle Drawer ReactFX"
    scene = new Scene {
      content = new BorderPane() {
        padding = Insets(10)
        top = new HBox(10) {
          padding = Insets(10)
          content = Seq(undo, redo)
        }
        center = canvas
      }
    }
  }
}

class CircleDrawerCanvasFX extends Canvas(400, 400) {
  var circles = Seq[CircleFX]()

  val diameter = new Button("Diameter...")
  val popup = new Popup()
  popup.content.add(diameter)

  val leftPressesToVoid: EventStream[javafx.scene.input.MouseEvent] = this.mousePresses
          .filter((e => e.isPrimaryButtonDown): (javafx.scene.input.MouseEvent => Boolean))
          .filter((e => getNearestCircleAt(e.getX, e.getY) == null): (javafx.scene.input.MouseEvent => Boolean))
  val addedCircles: EventStream[CircleFX] = leftPressesToVoid
          .map((e => new CircleFX(e.getX, e.getY)): (javafx.scene.input.MouseEvent => CircleFX))
          // appending hook lead to compile error so subscribe it is
          addedCircles.subscribe((c => addCircle(c)): (CircleFX => Unit))
  val hoveredCircles: EventStream[CircleFX] = this.mouseMoves
          .map((e => getNearestCircleAt(e.getX, e.getY)): (javafx.scene.input.MouseEvent => CircleFX))
  merge(addedCircles, hoveredCircles).subscribe((c => draw(c)): (CircleFX => Unit))

  this.mousePresses.subscribe((e => if (popup.isShowing) popup.hide()): (javafx.scene.input.MouseEvent => Unit))
  val rightPressesToCircle: EventStream[javafx.scene.input.MouseEvent] = this.mousePresses
          .filter((e => e.isPopupTrigger): (javafx.scene.input.MouseEvent => Boolean))
          .filter((e => getNearestCircleAt(e.getX, e.getY) != null): (javafx.scene.input.MouseEvent => Boolean))
  rightPressesToCircle.subscribe((e => popup.show(this, e.getScreenX, e.getScreenY)): (javafx.scene.input.MouseEvent => Unit))
  val selectedCircles: EventStream[CircleFX] = rightPressesToCircle
          .map((e => getNearestCircleAt(e.getX, e.getY)): (javafx.scene.input.MouseEvent => CircleFX))
  selectedCircles.emitOn(diameter.actions).subscribe((c => {popup.hide(); showDialog(c)}): (CircleFX => Unit))

  draw()

  def draw(hovered: CircleFX = null) = {
    val g = graphicsContext2D
    g.setFill(Color.WHITE)
    g.setStroke(Color.BLACK)
    g.fillRect(0, 0, width.value, height.value)
    g.strokeRect(0, 0, width.value, height.value)

    for (c <- circles) {
      val offset = c.d / 2
      if (c == hovered) {
        g.setFill(Color.LIGHTGRAY)
        g.fillOval(c.x-offset, c.y-offset, c.d, c.d)
      }
      g.strokeOval(c.x-offset, c.y-offset, c.d, c.d)
    }
  }

  def addCircle(circle: CircleFX) {
    circles = circles :+ circle
    addSnapshot()
  }

  def getNearestCircleAt(x: Double, y: Double) = {
    var circle: CircleFX = null
    var minDist = Double.MaxValue
    for (c <- circles) {
      val d = sqrt(pow(x-c.x, 2) + pow(y-c.y, 2))
      if (d <= c.d/2 && d < minDist) {
        circle = c
        minDist = d
      }
    }
    circle
  }

  def showDialog(selected: CircleFX) {
    val dialog = new Stage()
    val info = new Label(s"Adjust diameter of circle at (${selected.x}, ${selected.y})")
    val slider = new Slider(10, 50, selected.d)

    val index = circles.indexOf(selected)
    slider.value.addListener((v: ObservableValue[_ <: Number], o: Number, n: Number) => {
      circles = circles.updated(index, selected.copy(d=n.doubleValue))
      draw()
    })
    dialog.onCloseRequest = (e: WindowEvent) => addSnapshot()

    dialog.scene = new Scene() {
      content = new VBox(10) {
        padding = Insets(10)
        content = Seq(info, slider)
      }
    }
    dialog.show()
  }

  var history = Seq[Seq[CircleFX]](Seq())
  var historyCursor = 0

  def addSnapshot() = {
    history = history.take(historyCursor+1) :+ circles
    historyCursor += 1
  }

  def undo() = if (historyCursor > 0) {
    historyCursor -= 1
    circles = history(historyCursor)
    draw()
  }

  def redo() = if (historyCursor < history.size-1) {
    historyCursor += 1
    circles = history(historyCursor)
    draw()
  }

}

case class CircleFX(x: Double, y: Double, d: Double = 30)